mgr/dashboard: Introduce NVMe/TCP navigation

afreen23 opened the pull request:
Fixes https://tracker.ceph.com/issues/66346

- adds NVMe/TCP tab under Block nav
- adds overview page for NVMe/TCP nav
- overview page lists gateways
- add default error page when no nvmeof service running
- added unit tests
- fixes service page e2e test

Signed-off-by: Afreen <afreen23.git@gmail.com>

![image](https://github.com/ceph/ceph/assets/25664409/ff4cb386-7032-48cd-a980-31c751d86da8)

![Screenshot from 2024-06-22 20-42-54](https://github.com/ceph/ceph/assets/25664409/09385bd8-d7e5-452f-b897-773bb8645d0b)

[Screencast from 2024-06-24 17-05-17.webm](https://github.com/ceph/ceph/assets/25664409/306671a9-b6b1-4426-ae05-e3390961f2c7)


## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/block.module.ts:
```diff
@@ -198,6 +200,16 @@ const routes: Routes = [
+  {
+    path: 'nvmeof',
+    canActivate: [FeatureTogglesGuardService],
+    data: { breadcrumbs: 'Nvme-oF' },
```
I think the suggestion was to call it NVMe/TCP on one of the presentation so let's do it like it.
```suggestion
    data: { breadcrumbs: 'NVMe/TCP' },
```

nizamial09 reviewed (commented):
while we are at this, shall we add one navigation to the side bar as well inside the `Block`?

afreen23 commented:
jenkins test dashboard

afreen23 commented:
jenkins test make check

afreen23 commented:
jenkins test dashboard cephadm

afreen23 commented:
jenkins test dashboard

afreen23 commented:
Summary:

* rebased PR
* fixed e2e tests for cephadm for service page (missed in previous PR)

afreen23 commented:
jenkins test dashboard cephadm

afreen23 commented:
jenkins test api

afreen23 commented:
jenkins test make check

afreen23 commented:
jenkins test dashboard cephadm

afreen23 commented:
jenkins test dashboard

afreen23 commented:
jenkins test make check

afreen23 commented:
jenkins test api

afreen23 commented:
jenkins test dashboard cephadm

afreen23 commented:
Dahsboard tests failing due to https://tracker.ceph.com/issues/66506

afreen23 commented:
jenkins test api

afreen23 commented:
jenkins test api

afreen23 commented:
Summary of new changes:

- Based on Paul's feedback added service creation/deletion/edit from gateways itself. cc @pcuzner 
- Added help text for gateways

(waiting for new setup to add video rec)

afreen23 commented:
Summary of new changes:

   * Based on Paul's feedback added service creation/deletion/edit from gateways itself. cc @pcuzner
   * Added help text for gateways
   * Changed gateways text -> gateway group as thats the newly adopted entity in backend
   * Allow only one gateway group creation (will also block later from Administration -> services page, in a separate PR as some more enhancements needed there hence will send altogether)

Updated Screencast  cc @nizamial09 @cloudbehl

afreen23 commented:
jenkins test make check

afreen23 commented:
jenkins test dashboard

afreen23 commented:
jenkins test dashboard cephadm

afreen23 commented:
jenkins test make check arm64

afreen23 commented:
Summary of changes:

* updated screenshots and PR summary
* fallback page added when no nvmeof service running (no daemon is found)
* service form bugs fixed
* Added nvmeof permissions - crud support , required for upcoming PRs
* added a param in service form /create url to dynamically populate the service name. This is to improve user experience. We are creating service from nfs views as well, so will allow auto-populating the service name instead of clicking on selection. [Adapt nfs changed later]

Thanks

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/nvmeof.service.ts:
can you move the services under https://github.com/ceph/ceph/tree/main/src/pybind/mgr/dashboard/frontend/src/app/shared/api

afreen23 replied:
Done.
Linting errors are not reported as I checked now on my machine. Lets see if its now too.

afreen23 replied:
Ah! those were tox ones!, fixed now

nizamial09 reviewed (commented):
there are some lint failures too

nizamial09 reviewed (approved):
Thank you @afreen23 functionally this looks great!

nizamial09 commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -203,7 +215,7 @@ def io_stats(self, nqn: str, nsid: str):
             parameters={
                 "nqn": Param(str, "NVMeoF subsystem NQN"),
-                "rbd_pool": Param(str, "RBD pool name"),
+                "rbd_potoxol": Param(str, "RBD pool name"),
```
is this intentional?

nizamial09 commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -1,15 +1,27 @@
         handle_nvmeof_error, map_collection, map_model
-except ImportError:
+except ImportError as e:
+    logging.error("Failed to import NVMeoFClient and related components: %s", e)
```
```suggestion
    logger.error("Failed to import NVMeoFClient and related components: %s", e)
```

nizamial09 commented on src/pybind/mgr/dashboard/controllers/nvmeof.py:
```diff
@@ -1,15 +1,27 @@
+from . import APIDoc, APIRouter, UIRouter, Endpoint, \
+    EndpointDoc, Param, ReadPermission, BaseController, RESTController
+
+logger = logging.getLogger()
```
so the name of the module/file will be used
```suggestion
logger = logging.getLogger(__name__)
```

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/block/block.module.ts:
```diff
@@ -159,7 +161,7 @@ const routes: Routes = [
         button_name: $localize`Configure Block Mirroring`,
         button_title: $localize`This will create \"rbd-mirror\" service and a replicated Block pool`,
-        component: 'Block Mirroring',
+        component: 'Default Pool',
```
why this one?

nizamial09 reviewed (approved):
thank you @afreen23

cloudbehl commented:
1. When No RBD was present and I created the RBD and came back to NVME OF page , at that moment it gave me "NO page found" Error.
2. When we create first service and the daemon are not created or gets into error stage. At that point NVMe of page gives
504 error and user doesn't get idea of what to do.


![Screenshot from 2024-06-28 18-59-07](https://github.com/ceph/ceph/assets/6695156/0432a99f-33dd-4f85-b6b7-eedc31db87ce)
![Screenshot from 2024-06-28 19-03-41](https://github.com/ceph/ceph/assets/6695156/88276f24-cd5e-40f5-bbd1-f8df50def7c1)


While testing NVMe, I was not able to deploy daemon, it was getting into error stage. 
![Screenshot from 2024-06-28 19-05-37](https://github.com/ceph/ceph/assets/6695156/c46c7f8e-a198-4beb-b46c-102afed32bba)

afreen23 commented:
@nizamial09 thanks, I have updated changes, some were done mistakenly previously so reverted.

@cloudbehl thanks

1. For no RBD, I am correctly seeing the default page of Nvmeof
2.  That seems like error in fetching gateways as you mentioned daemon was in error. We  have `handle_nvmeof_error` dec for gateway API which throws such dashboard exceptions. cc @nizamial09 for he wrote these. can we improve here ?
3. This one looks like resource issue, try <= 8G ram/ 3 cpus

cloudbehl commented:
@afreen23 

1) For me it happened when I created the service and after creating immediately I came back to NVMe page. It showed me " No page found" but that can be corner case as well. 
2. Can we improve the notification message for it to make it better.

nizamial09 commented:
> 2\. That seems like error in fetching gateways as you mentioned daemon was in error. We  have `handle_nvmeof_error` dec for gateway API which throws such dashboard exceptions. cc @nizamial09 for he wrote these. can we improve here ?

the exception comes directly from the grpc when the requested resource is not there which in this case seems correct since the nvmf daemon is in error. the decorator catches all types of error and raises the exception with the error details and it looks like that's all the grpc is sending to us. So probably if all the nvmeof daemon is in error, we should hide the page with the error view saying why its in error or something like that. But I'd say let's do that separately. https://tracker.ceph.com/issues/66841

nizamial09 commented:
> 2\. Can we improve the notification message for it to make it better.

it is an exception raised from the back-end and simply just thrown in the UI. It'd be easier to hide the view and throw the exception in the error view rather than showing it in a notification. The same behaviour we have in the NFS page (or maybe others) when orch is not configured or other issues...

afreen23 commented:
> @afreen23
> 
>     1. For me it happened when I created the service and after creating immediately I came back to NVMe page. It showed me " No page found" but that can be corner case as well.
> 
> 
okay that is valid since daemon is in error and for that I will make chnages with https://tracker.ceph.com/issues/66841
but nothing related to "no RBD" case I noticed.
>     2. Can we improve the notification message for it to make it better.
lets do the same way as NFS which @nizamial09 mentioned

cc @cloudbehl

afreen23 commented:
jenkins test api

afreen23 commented:
jenkins test api

afreen23 commented:
jenkins test make check

afreen23 commented:
jenkins test make check arm64

afreen23 commented:
jenkins test make check

afreen23 commented:
jenkins test make check

afreen23 commented:
jenkins test make check

afreen23 commented:
jenkins test make check
